### PR TITLE
Decide on just one way to specify array dimensions

### PR DESCRIPTION
**Pull request by henrikt-ma**
Opening PR according to road map item scheduled at last web meeting.

Marking as draft PR, as I understand that diverging from the Modelica grammar is controversial.  My reasoning is that just settling for one of the ways to define array dimensions is disruptive enough to make an opportunity for also fixing the issue with the non-intuitiveness of Modelica array dimension concatenation.

This PR builds upon https://github.com/modelica/ModelicaSpecification/pull/2465, so it should remain in draft state at least until that has been merged.

### CODE REVIEW

**Comment by henrikt-ma** [comment]
Ouch.  It seems like a mistake to have based this PR on the unmerged https://github.com/modelica/ModelicaSpecification/pull/2465.  I'm afraid this means that we'll need to rebase this PR once https://github.com/modelica/ModelicaSpecification/pull/2465 has been merged, but I have no idea what that would do to the conversation…

**Comment by henrikt-ma** [comment]
> Ouch. It seems like a mistake to have based this PR on the unmerged #2465. I'm afraid this means that we'll need to rebase this PR once #2465 has been merged, but I have no idea what that would do to the conversation…

I've now rebased on _MCP/0031_, and — voila! — all the irrelevant comments from #2465 is gone.  However, very tragically, all the discussion we've had so far on the current matter is also gone.  I wonder where it is now, and whether there is a way to get it back…

**Comment by henrikt-ma** [comment]
Intermediate observation after resetting the branch to where it was before the rebase on _MCP/0031_: All comments from #2465 are gone!  I don't understand this, but will continue to merge with _MCP/0031_ anyway.

**Comment by olivleno** [comment]
Consequences of not allowing type aliases for variable declarations:
`Modelica.SIunits.Pressure p;`
`Real p (unit="Pa", displayUnit="bar", quantity="Pressure", min=0, nominal=1e5);`

Consequence for records:
```
record Point
    Real[3] x(unit="m", displayUnit="mm", quantity="Length", nominal=1);
end Point;
Point p;
p.x[1] = 0;
```
Consequences from tool vendor perspective:

- Type aliases force tools to handle merging of modifications across type hierarchies, which makes it much hard to parse and handle by a tool.
- Depends on what is allowed
`Modelica.SIunits.Pressure p(min=1);`
`Real p (unit="Pa", displayUnit="bar", quantity="Pressure", min=1, nominal=1e5);`
One could disallow nested modifications.

Next meeting:
- Continue the discussion trying to get Kai in the call to hear his opinion as a non-modelica tool vendor.

**Comment by olivleno** [comment]
**Discussion:**
Kai: Okay to have type aliases. Not too complicated.
Gerd: But how about modifications on types like `Modelica.SIunits.Pressure p(min=1);`
Henrik: This can be understood as a different default.

**Type Aliases for scalars:**
Option 1: No type aliases and no default attributes (votes=0)
```
   Real x(quantity="Length", unit="m", displayUnit="m", min=-INF, max=INF, start=0, fixed=false, nominal=1, unbounded=false, stateSelect=StateSelect.default);
   Real xx(quantity="Length", unit="m", displayUnit="m", min=-INF, max=INF, start=0, fixed=false, nominal=1e6, unbounded=false, stateSelect=StateSelect.default);
```
Option 2: No type aliases with default attributes (votes=2)
```
   Real x(quantity="Length", unit="m", displayUnit="m");
   Real xx(quantity="Length", unit="m", displayUnit="m", nominal=1e6);
```
Option 3: Type aliases with no default attributes (votes=1)
```
   type Length = Real (quantity="Length", unit="m", displayUnit="m", min=-INF, max=INF, start=0, fixed=false, nominal=1, unbounded=false, stateSelect=StateSelect.default);
   Length x(min=0, start=1);
   type BigLength = Length(nominal=1e6);
   BigLength xx;
```
Option 4: Type aliases with default attributes (votes=4)
```
   type Length = Real (quantity="Length", unit="m", displayUnit="m");
   Length x(min=0, start=1);
   type BigLength = Length(nominal=1e6);
```
**Result: Option 4**

**Comment by olivleno** [comment]
**Type Aliases for records:**
Option 1:
```
record Point
    Real[3] x(unit="m", displayUnit="m", quantity="Length", nominal=1, min=0);
end Point;
Point p;
p.x[1] = 0;
```
Option 2:
```
type Length = Real (unit="m", displayUnit="m", quantity="Length", nominal=1, start=0);
record Point
    Length[3] x(min={0,0,1});
end Point;
Point p(x(start={0,0,1});

Point p={x(start={0,0,1});
p.x[1] = 0;
```
Option 3:
```
type Point = Real[3] (min={0,0,1});
Point p;
p[1] = 0;
```

**Comment by henrikt-ma** [comment]
The discussion above on allowed forms of type aliases is mainly out of topic and should be continued here: https://github.com/modelica/ModelicaSpecification/pull/2555

Corresponding design document: https://github.com/modelica/ModelicaSpecification/blob/MCP/0031%2Btype-aliases/RationaleMCP/0031/type-aliases.md

**Comment by henrikt-ma** [comment]
The comment above from 2020-04-28 still applies: There's no point discussing how to specify array dimensions until we have settled the question about allowing array type aliases at all; without array type aliases, there is no potentially confusing order of dimensions for this PR to resolve, and we can just go with what looks most natural to us.

**Comment by olivleno** [comment]
Web Meeting: Hans, Henrik, Gerd, Martin, Oliver

Proposal to resolve the confusion as shown in the initial example from above:
```
model ModelicaAppendDimensions
  //type Arr = Real[3, 4];
  type Arr = [3, 4]Real;
  [1, 2]Arr x = fill(0.0, 1, 2, 3, 4); /* Note: [k, l]([m, n]Real) is the same as [k, l, m, n]Real. */
end ModelicaAppendDimensions;
```

Henrik: This could be a way to make it more consistent.

Hans: This is very different. Don't see this as an issue. It causes more problems to solve a rare case by introducing something new that is very different.

Henrik: Just having one way does not solve the problem. I consider it not as minor problem.
Remove confusion by disallowing type aliases for arrays, but once we decide later to introduce them we'll run into the same problem again.

Gerd: On first sight it looks ugly, but could become familiar as you get used to it.

Martin: Would be easier to digest if it becomes part of Modelica first.

Henrik: Long-term this could bubble up to Modelica, but I don't see it prerequisite. But having three ways will not help.
We could introduce the new way an still support one of the old ways for backward compatibility.

Hans: It's more a matter of nesting that introduces the confusion, buy having the type in front of the component declaration.

Oliver: We have the goal to stick to Modelica as much as possible. Solving this issue in flat Modelica seems to me to be not a primary concern, therefore I'd like to delay this problem.

Henrik: If we disallow type aliases and only allow array dimension after component names, the we could fix it later.

Gerd: Type aliases are rather a feature for hand written flat Modelica. Disallowing them for now would be okay.

Proposal:
Disallow  type aliases and only allow array dimensions after component names.

Poll
Accept: Henrik, Gerd, Martin, Hans, Oliver

Conclusion:
Accept the proposal.
Update the current PR [Henrik]

**Comment by HansOlsson** [comment]
Just clarifying my nesting comment above.

One issue is that in Modelica the order is somewhat odd for corner cases:
```
   type A34=Real[3,4]
   A34 x[1,2]=ones(1,2,3,4);
```
My point is that we have the same order if we use record members (in Modelica) - it's just that we don't even realize it:
```
   record R34
      Real z[3,4];
   end R34;
   R34 r[1,2](z=ones(1,2,3,4));
```
(Pascal avoids that by using a different syntax: `var r[1,2] : R34`)

**Comment by henrikt-ma** [comment]
Removing _Draft_ state after having implemented the decision in https://github.com/modelica/ModelicaSpecification/pull/2468#issuecomment-1112953044.

**Review by gkurzbach — APPROVED** [review]
I'm fine with it.
